scene/3d/physics_body_3d.cpp

```diff
@@ -1171,12 +1167,12 @@ bool CharacterBody3D::move_and_slide() {
 
 		sliding_enabled = true;
 
-		if (!found_collision || motion == Vector3()) {
+		if (!collided || motion.is_equal_approx(Vector3())) {
 			break;
 		}
 	}
 
-	if (was_on_floor && snap != Vector3()) {
+	if (was_on_floor && !snap.is_equal_approx(Vector3())) {
```

Comment:
We added `&& !on_floor` in a previous PR, you can add it, or I will do it later as I need to work on this file